{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Pipeline to delimit ICE"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This notebook has been cloned from https://github.com/gem-pasteur/Macsyfinder_models/blob/master/models/Conjugation/Tutorial_ICE.ipynb?short_path=6cda664 in order to confirm my previous findings!\n",
    "\n",
    "In this notebook, we'll find ICE and delimit them in four Lactobacillus mudanjiangesis species\n",
    "\n",
    "1. First we'll get the genomes\n",
    "3. We'll build the core genome\n",
    "2. We'll detect the conjugative system in the genomes\n",
    "4. We'll identify the core-genes flanking the conjugative system (defining a spot)\n",
    "5. We'll build the pan-genome of the spot\n",
    "6. We'll make a visual representation of the spot with gene family frequency to delimit the ICE."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<p><div class=\"lev1\"><a href=\"#Data\"><span class=\"toc-item-num\">2&nbsp;&nbsp;</span>Data</a></div><div class=\"lev2\"><a href=\"#Download-fasta-sequences\"><span class=\"toc-item-num\">2.1&nbsp;&nbsp;</span>Download fasta sequences</a></div><div class=\"lev2\"><a href=\"#Make-CDS-annotation-with-Prokka\"><span class=\"toc-item-num\">2.2&nbsp;&nbsp;</span>Make CDS annotation with Prokka</a></div><div class=\"lev1\"><a href=\"#Build-Core-genomes\"><span class=\"toc-item-num\">3&nbsp;&nbsp;</span>Build Core genomes</a></div><div class=\"lev1\"><a href=\"#Identification-of-Conjugative-systems\"><span class=\"toc-item-num\">4&nbsp;&nbsp;</span>Identification of Conjugative systems</a></div><div class=\"lev1\"><a href=\"#Identify-the-flanking-core-genes:-finding-a-spot\"><span class=\"toc-item-num\">5&nbsp;&nbsp;</span>Identify the flanking core genes: finding a spot</a></div><div class=\"lev1\"><a href=\"#Pan-Genome-of-the-spot\"><span class=\"toc-item-num\">6&nbsp;&nbsp;</span>Pan-Genome of the spot</a></div><div class=\"lev1\"><a href=\"#Format-data-to-plot\"><span class=\"toc-item-num\">7&nbsp;&nbsp;</span>Format data to plot</a></div><div class=\"lev2\"><a href=\"#Make-a-table-for-each-ICE\"><span class=\"toc-item-num\">7.1&nbsp;&nbsp;</span>Make a table for each ICE</a></div><div class=\"lev1\"><a href=\"#Plot-the-spot\"><span class=\"toc-item-num\">8&nbsp;&nbsp;</span>Plot the spot</a></div>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Approximate time ot run this notebook: ~15-20 minutes.\n",
    "\n",
    "**NB**: The pipeline here works because the 3 ICEs we will found are in the same spot. But often, ICEs are found in different spots, and it implies to deal with that information to not mix everything."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's transfer the necessary files to the right folder"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "mkdir out_ICE_analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "mkdir out_ICE_analysis/in"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "!cp ~/Desktop/clade1s/*gff out_ICE_analysis/in"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Build Core genomes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "With [Roary](https://sanger-pathogens.github.io/Roary/)\n",
    "\n",
    "If roary was not yet installed:\n",
    "\n",
    "```bash\n",
    "$ sudo apt-get install bedtools cd-hit blast mcl parallel prank mafft fasttree cpanm\n",
    "$ sudo cpanm -f Bio::Roary\n",
    "```\n",
    "\n",
    "Move all `.gff` files in the same folder so Roary can work with them"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "mkdir out_ICE_analysis/Pangenome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "!roary -p 8 -f out_ICE_analysis/Pangenome/ out_ICE_analysis/in/*.gff > out_ICE_analysis/LOG_roary 2>&1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Get core-genes\n",
    "\n",
    "**BEWARE** to change the random number generated by `roary` (here `_1536417924`)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "!query_pan_genome -g out_ICE_analysis/Pangenome/_1536417924/clustered_proteins -a intersection out_ICE_analysis/in/*.gff"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "mv pan_genome_results out_ICE_analysis/core_genome.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "group_319: BIHEEGNE_00936\tIGPPGNFN_01861\tMDHGFNIF_01370\tDOGOELCO_02343\r\n",
      "group_3250: BIHEEGNE_01602\tIGPPGNFN_03400\tMDHGFNIF_01268\tDOGOELCO_00808\r\n",
      "group_3083: BIHEEGNE_00153\tIGPPGNFN_01744\tMDHGFNIF_01121\tDOGOELCO_01936\r\n",
      "group_2531: BIHEEGNE_02538\tIGPPGNFN_00185\tMDHGFNIF_00222\tDOGOELCO_01255\r\n",
      "group_3745: BIHEEGNE_02627\tIGPPGNFN_00318\tMDHGFNIF_00090\tDOGOELCO_01175\r\n",
      "group_3031: BIHEEGNE_02877\tIGPPGNFN_00064\tMDHGFNIF_00343\tDOGOELCO_01364\r\n",
      "group_4136: BIHEEGNE_00169\tIGPPGNFN_01728\tMDHGFNIF_01106\tDOGOELCO_01921\r\n",
      "group_4017: BIHEEGNE_03328\tIGPPGNFN_01386\tMDHGFNIF_03238\tDOGOELCO_02910\r\n",
      "group_4229: BIHEEGNE_01884\tIGPPGNFN_00660\tMDHGFNIF_02197\tDOGOELCO_02994\r\n",
      "group_2203: BIHEEGNE_00302\tIGPPGNFN_01656\tMDHGFNIF_00972\tDOGOELCO_01786\r\n"
     ]
    }
   ],
   "source": [
    "!tail out_ICE_analysis/core_genome.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2786 out_ICE_analysis/core_genome.txt\r\n"
     ]
    }
   ],
   "source": [
    "!wc -l out_ICE_analysis/core_genome.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Remove first columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "!awk -F\":\" '{print $2}' out_ICE_analysis/core_genome.txt > out_ICE_analysis/core_genome_2.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Identification of Conjugative systems"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "I downloaded the [module CONJScan](https://minhaskamal.github.io/DownGit/#/home?url=https://github.com/gem-pasteur/Macsyfinder_models/tree/master/Data/Conjugation) with model and profiles for conjugative systems to use with macsyfinder. I put it under a Conjugation folder and rename it \"Models\".\n",
    "\n",
    "I already ran this analysis and copied the results to the folder Conjugation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/swuyts/serverUA/sander/mudan/13_plasmids/out_ICE_analysis\n"
     ]
    }
   ],
   "source": [
    "cd out_ICE_analysis/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "   0    0    0 Conjugation/AMB-F197_typeB/macsyfinder.report\n",
      "   0    0    0 Conjugation/AMB-F197_typeC/macsyfinder.report\n",
      "   0    0    0 Conjugation/AMB-F197_typeF/macsyfinder.report\n",
      "   0    0    0 Conjugation/AMB-F197_typeFA/macsyfinder.report\n",
      "   0    0    0 Conjugation/AMB-F197_typeFATA/macsyfinder.report\n",
      "   0    0    0 Conjugation/AMB-F197_typeG/macsyfinder.report\n",
      "   0    0    0 Conjugation/AMB-F197_typeI/macsyfinder.report\n",
      "   0    0    0 Conjugation/AMB-F197_typeT/macsyfinder.report\n",
      "   0    0    0 Conjugation/AMB-F209_typeB/macsyfinder.report\n",
      "   0    0    0 Conjugation/AMB-F209_typeC/macsyfinder.report\n",
      "   0    0    0 Conjugation/AMB-F209_typeF/macsyfinder.report\n",
      "   0    0    0 Conjugation/AMB-F209_typeFA/macsyfinder.report\n",
      "  13  208 2119 Conjugation/AMB-F209_typeFATA/macsyfinder.report\n",
      "   0    0    0 Conjugation/AMB-F209_typeG/macsyfinder.report\n",
      "   0    0    0 Conjugation/AMB-F209_typeI/macsyfinder.report\n",
      "   0    0    0 Conjugation/AMB-F209_typeT/macsyfinder.report\n",
      "   0    0    0 Conjugation/AMB-F249_typeB/macsyfinder.report\n",
      "   0    0    0 Conjugation/AMB-F249_typeC/macsyfinder.report\n",
      "   0    0    0 Conjugation/AMB-F249_typeF/macsyfinder.report\n",
      "   0    0    0 Conjugation/AMB-F249_typeFA/macsyfinder.report\n",
      "  14  224 2274 Conjugation/AMB-F249_typeFATA/macsyfinder.report\n",
      "   0    0    0 Conjugation/AMB-F249_typeG/macsyfinder.report\n",
      "   0    0    0 Conjugation/AMB-F249_typeI/macsyfinder.report\n",
      "   0    0    0 Conjugation/AMB-F249_typeT/macsyfinder.report\n",
      "   0    0    0 Conjugation/L-Mudanjiangensis-DSM-28402_typeB/macsyfinder.report\n",
      "   0    0    0 Conjugation/L-Mudanjiangensis-DSM-28402_typeC/macsyfinder.report\n",
      "   0    0    0 Conjugation/L-Mudanjiangensis-DSM-28402_typeF/macsyfinder.report\n",
      "   0    0    0 Conjugation/L-Mudanjiangensis-DSM-28402_typeFA/macsyfinder.report\n",
      "   8  128 1321 Conjugation/L-Mudanjiangensis-DSM-28402_typeFATA/macsyfinder.report\n",
      "   0    0    0 Conjugation/L-Mudanjiangensis-DSM-28402_typeG/macsyfinder.report\n",
      "   0    0    0 Conjugation/L-Mudanjiangensis-DSM-28402_typeI/macsyfinder.report\n",
      "   0    0    0 Conjugation/L-Mudanjiangensis-DSM-28402_typeT/macsyfinder.report\n",
      "  35  560 5714 total\n"
     ]
    }
   ],
   "source": [
    "!wc Conjugation/*/macsyfinder.report"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We have 3 ICEs of type FATA in strain AMB-F209, AMB-F249 and DSM28402"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "!awk 'NR==1||FNR>1' Conjugation/*/macsyfinder.report > Conj_results"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Identify the flanking core genes: finding a spot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "conj = pd.read_table(\"Conj_results\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "core = pd.read_table(\"core_genome_2.txt\", header=None, names=[\"BIHEEGNE\", \"IGPPGNFN\", \"MDHGFNIF\", \"DOGOELCO\"])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Reformat Macsyfinder output, replace UserReplicon by the name of the replicon. (I may have missed an option)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "conj[\"Replicon_name\"] = conj[\"#Hit_Id\"].apply(lambda x: x[:8])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "conj.System_Id.replace(\"UserReplicon\", conj.Replicon_name, regex=True, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#Hit_Id</th>\n",
       "      <th>Replicon_name</th>\n",
       "      <th>Position</th>\n",
       "      <th>Sequence_length</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Reference_system</th>\n",
       "      <th>Predicted_system</th>\n",
       "      <th>System_Id</th>\n",
       "      <th>System_status</th>\n",
       "      <th>Gene_status</th>\n",
       "      <th>i-evalue</th>\n",
       "      <th>Score</th>\n",
       "      <th>Profile_coverage</th>\n",
       "      <th>Sequence_coverage</th>\n",
       "      <th>Begin_match</th>\n",
       "      <th>End_match</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>IGPPGNFN_02837</td>\n",
       "      <td>IGPPGNFN</td>\n",
       "      <td>2799</td>\n",
       "      <td>687</td>\n",
       "      <td>MOBQ</td>\n",
       "      <td>CONJ</td>\n",
       "      <td>typeFATA</td>\n",
       "      <td>IGPPGNFN_typeFATA_1</td>\n",
       "      <td>single_locus</td>\n",
       "      <td>mandatory</td>\n",
       "      <td>7.100000e-77</td>\n",
       "      <td>255.7</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.335</td>\n",
       "      <td>1</td>\n",
       "      <td>230</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>IGPPGNFN_02841</td>\n",
       "      <td>IGPPGNFN</td>\n",
       "      <td>2803</td>\n",
       "      <td>120</td>\n",
       "      <td>FATA_trsC</td>\n",
       "      <td>typeFATA</td>\n",
       "      <td>typeFATA</td>\n",
       "      <td>IGPPGNFN_typeFATA_1</td>\n",
       "      <td>single_locus</td>\n",
       "      <td>accessory</td>\n",
       "      <td>1.800000e-49</td>\n",
       "      <td>164.8</td>\n",
       "      <td>0.874</td>\n",
       "      <td>0.942</td>\n",
       "      <td>1</td>\n",
       "      <td>113</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>IGPPGNFN_02842</td>\n",
       "      <td>IGPPGNFN</td>\n",
       "      <td>2804</td>\n",
       "      <td>219</td>\n",
       "      <td>FATA_trsD</td>\n",
       "      <td>typeFATA</td>\n",
       "      <td>typeFATA</td>\n",
       "      <td>IGPPGNFN_typeFATA_1</td>\n",
       "      <td>single_locus</td>\n",
       "      <td>accessory</td>\n",
       "      <td>3.000000e-108</td>\n",
       "      <td>358.2</td>\n",
       "      <td>0.960</td>\n",
       "      <td>1.000</td>\n",
       "      <td>1</td>\n",
       "      <td>219</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>IGPPGNFN_02843</td>\n",
       "      <td>IGPPGNFN</td>\n",
       "      <td>2805</td>\n",
       "      <td>672</td>\n",
       "      <td>virb4</td>\n",
       "      <td>CONJ</td>\n",
       "      <td>typeFATA</td>\n",
       "      <td>IGPPGNFN_typeFATA_1</td>\n",
       "      <td>single_locus</td>\n",
       "      <td>mandatory</td>\n",
       "      <td>1.800000e-98</td>\n",
       "      <td>328.7</td>\n",
       "      <td>0.587</td>\n",
       "      <td>0.891</td>\n",
       "      <td>62</td>\n",
       "      <td>660</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>IGPPGNFN_02848</td>\n",
       "      <td>IGPPGNFN</td>\n",
       "      <td>2810</td>\n",
       "      <td>156</td>\n",
       "      <td>FATA_trsJ</td>\n",
       "      <td>typeFATA</td>\n",
       "      <td>typeFATA</td>\n",
       "      <td>IGPPGNFN_typeFATA_1</td>\n",
       "      <td>single_locus</td>\n",
       "      <td>accessory</td>\n",
       "      <td>1.300000e-05</td>\n",
       "      <td>22.9</td>\n",
       "      <td>0.857</td>\n",
       "      <td>0.776</td>\n",
       "      <td>20</td>\n",
       "      <td>140</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          #Hit_Id Replicon_name  Position  Sequence_length       Gene  \\\n",
       "0  IGPPGNFN_02837      IGPPGNFN      2799              687       MOBQ   \n",
       "1  IGPPGNFN_02841      IGPPGNFN      2803              120  FATA_trsC   \n",
       "2  IGPPGNFN_02842      IGPPGNFN      2804              219  FATA_trsD   \n",
       "3  IGPPGNFN_02843      IGPPGNFN      2805              672      virb4   \n",
       "4  IGPPGNFN_02848      IGPPGNFN      2810              156  FATA_trsJ   \n",
       "\n",
       "  Reference_system  Predicted_system            System_Id System_status  \\\n",
       "0             CONJ  typeFATA          IGPPGNFN_typeFATA_1  single_locus   \n",
       "1         typeFATA  typeFATA          IGPPGNFN_typeFATA_1  single_locus   \n",
       "2         typeFATA  typeFATA          IGPPGNFN_typeFATA_1  single_locus   \n",
       "3             CONJ  typeFATA          IGPPGNFN_typeFATA_1  single_locus   \n",
       "4         typeFATA  typeFATA          IGPPGNFN_typeFATA_1  single_locus   \n",
       "\n",
       "  Gene_status       i-evalue  Score  Profile_coverage  Sequence_coverage  \\\n",
       "0   mandatory   7.100000e-77  255.7             1.000              0.335   \n",
       "1   accessory   1.800000e-49  164.8             0.874              0.942   \n",
       "2   accessory  3.000000e-108  358.2             0.960              1.000   \n",
       "3   mandatory   1.800000e-98  328.7             0.587              0.891   \n",
       "4   accessory   1.300000e-05   22.9             0.857              0.776   \n",
       "\n",
       "   Begin_match  End_match  \n",
       "0            1        230  \n",
       "1            1        113  \n",
       "2            1        219  \n",
       "3           62        660  \n",
       "4           20        140  "
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "conj.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To identify the flanking core genes, I sort the column in the table `core` corresponding to the genome where I have an ICE, and find the index where I would insert any genes of that ICE."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "             BIHEEGNE        IGPPGNFN        MDHGFNIF        DOGOELCO\n",
      "2143   BIHEEGNE_02430  IGPPGNFN_03561  MDHGFNIF_01563  DOGOELCO_00002\n",
      "1624   BIHEEGNE_02427  IGPPGNFN_03564  MDHGFNIF_01566  DOGOELCO_00005\n",
      "176    BIHEEGNE_02422  IGPPGNFN_03569  MDHGFNIF_01571  DOGOELCO_00010\n",
      "1736   BIHEEGNE_02421  IGPPGNFN_03570  MDHGFNIF_01572  DOGOELCO_00011\n",
      "1501   BIHEEGNE_02416  IGPPGNFN_03575  MDHGFNIF_01577  DOGOELCO_00016\n",
      "[2736 2737]\n",
      "             BIHEEGNE        IGPPGNFN        MDHGFNIF        DOGOELCO\n",
      "214    BIHEEGNE_02934  IGPPGNFN_00001  MDHGFNIF_00406  DOGOELCO_01427\n",
      "1178   BIHEEGNE_02933  IGPPGNFN_00002  MDHGFNIF_00405  DOGOELCO_01426\n",
      "2761   BIHEEGNE_02932  IGPPGNFN_00003  MDHGFNIF_00404  DOGOELCO_01425\n",
      "241    BIHEEGNE_02931  IGPPGNFN_00004  MDHGFNIF_00403  DOGOELCO_01424\n",
      "2598   BIHEEGNE_02930  IGPPGNFN_00005  MDHGFNIF_00402  DOGOELCO_01423\n",
      "[2242 2243]\n",
      "             BIHEEGNE        IGPPGNFN        MDHGFNIF        DOGOELCO\n",
      "214    BIHEEGNE_02934  IGPPGNFN_00001  MDHGFNIF_00406  DOGOELCO_01427\n",
      "1178   BIHEEGNE_02933  IGPPGNFN_00002  MDHGFNIF_00405  DOGOELCO_01426\n",
      "2761   BIHEEGNE_02932  IGPPGNFN_00003  MDHGFNIF_00404  DOGOELCO_01425\n",
      "241    BIHEEGNE_02931  IGPPGNFN_00004  MDHGFNIF_00403  DOGOELCO_01424\n",
      "2598   BIHEEGNE_02930  IGPPGNFN_00005  MDHGFNIF_00402  DOGOELCO_01423\n",
      "[2363]\n",
      "             BIHEEGNE        IGPPGNFN        MDHGFNIF        DOGOELCO\n",
      "1810   BIHEEGNE_02825  IGPPGNFN_00407  MDHGFNIF_00001  DOGOELCO_01086\n",
      "885    BIHEEGNE_02826  IGPPGNFN_00406  MDHGFNIF_00002  DOGOELCO_01087\n",
      "1905   BIHEEGNE_02827  IGPPGNFN_00405  MDHGFNIF_00003  DOGOELCO_01088\n",
      "2419   BIHEEGNE_02828  IGPPGNFN_00404  MDHGFNIF_00004  DOGOELCO_01089\n",
      "721    BIHEEGNE_02829  IGPPGNFN_00403  MDHGFNIF_00005  DOGOELCO_01090\n",
      "[2707 2708 2709 2710 2740]\n"
     ]
    }
   ],
   "source": [
    "flanking_core = {}\n",
    "for ice in conj.groupby(\"System_Id\"):\n",
    "    data = ice[1]\n",
    "    replicon = data.Replicon_name.unique()[0]\n",
    "    core.sort_values(replicon, inplace=True)\n",
    "    print(core.head())\n",
    "    spots = np.unique([core[replicon].searchsorted(gene)[0] for gene in data[\"#Hit_Id\"]])\n",
    "    print(spots)\n",
    "  #  assert len(spots)==1 # make sure that all genes of the conjugative system insert at the same spot\n",
    "    flanking_core[ice[0]] = core.iloc[spots[0]-1:spots[0]+1].index\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This analysis did not work out...\n",
    "\n",
    "Maybe try in R?\n"
   ]
  }
 ],
 "metadata": {
  "hide_input": false,
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.5"
  },
  "toc": {
   "toc_cell": false,
   "toc_number_sections": true,
   "toc_section_display": "block",
   "toc_threshold": 6,
   "toc_window_display": true
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
